### PR TITLE
Generating offline report on HPC cluster

### LINKED ISSUE #2204: Replacing "gs" strings to run ML workflow on Stellar cluster
Machine learning (ML) workflow consists of several stages, which involve running numerical model (FV3GFS) and traning ML model. Perfroming ML training, I realized that some of "gs"-strings ("gs" - Google Cloud) in the catalog.yaml file 
https://github.com/ai2cm/fv3net/blob/3b1ff5c9378f6a70c8b760e4f02df373d32de780/external/vcm/vcm/catalog.yaml#L363

need to be replaced with the pathes to my local home directory e.g.

`urlpath: "/home/mr7417/ML_workflow/c48_fv3config/data/vcm_catalog/c12.zarr/"`

In order to remain consistent with the existing structure of the code (i.e. the user should be able to run the ML workflow on both Google cloud platfrom and HPC cluster), some more elaborate (or elegant) code changes need to be introduced.

### PR DESCRIPTION
**Pull request by mrudko**
Generating offline report consists of two steps: 1) generating the offline diagnostics datasets, $R^2$ and bias metrics for a given ML model, 2) compiling HTML report with tables and figures. The first step inevitably involves loading grid information, and, as a consequence, interacting with `catalog.yaml` file. The existing `catalog.yaml` file specifies the location of grid dataset with the cloud-style path strings. This apparently obtrudes creating the offline report on HPC cluster. This PR introduces the code changes, which intend to extend the usage of `catalog.yaml` file on HPC cluster.

Added public API:
- Added `catalog_path` argument to the CLI in the `compute.py` script

Refactored public API:
- Refactored `get_prediction` function

Significant internal changes:
- Added `catalog_path` input argument to `batches_from_mapper` function
- Refactored `_coarsen_transform` function (`compute.py` script)
- Refactored `load_grid_info` and `load_grid_info_fv3` functions (`_helpers.py`)

- [x] Tests added
- Added test `test_load_grid_info_from_custom_catalog` to the `test_helpers.py` script.

Resolves #2204 (partially)

### CODE REVIEW

File: workflows/diagnostics/fv3net/diagnostics/offline/_helpers.py
Comment (spencerkclark) [inline workflows/diagnostics/fv3net/diagnostics/offline/_helpers.py:15]:
```suggestion
```

**Review by spencerkclark — COMMENTED** [review]
@mrudko now that #2219 has been merged (thanks!), could you go ahead and update this PR to remove the changes to the files in the `loaders` package?  

I'll think a little more about whether there is an easy way to test aspects of this change.

**Review by spencerkclark — COMMENTED** [review]
Great, thanks for updating this with the new test @mrudko.  Could you merge in the master branch, address the subsequent merge conflicts, and then push again?  I think we should eventually see no changes to the `loaders` package (part of the reason we see them now is that your current branch is behind the master branch, so we need to merge the master branch in to catch up).

This is what I mean by merging in the master branch:
```
$ git checkout master
$ git pull origin master
$ git checkout fix-2204_1
$ git merge master

< ...address the merge conflicts... >

$ git push origin fix-2204_1
```

File: external/loaders/tests/test__batch.py
Comment (spencerkclark) [inline external/loaders/tests/test__batch.py:20]:
```suggestion

```

File: external/loaders/loaders/batches/_batch.py
Comment (spencerkclark) [inline external/loaders/loaders/batches/_batch.py:151]: Not strictly necessary, but we might as well reduce the diff.
```suggestion
```

File: external/loaders/loaders/batches/_batch.py
Comment (spencerkclark) [inline external/loaders/loaders/batches/_batch.py:180]: Not strictly necessary, but we might as well reduce the diff.
```suggestion
```

**Review by spencerkclark — COMMENTED** [review]
Thanks @mrudko -- it looks like there was a remaining merge conflict.

**Comment by mrudko** [comment]
I will fix it within half an hour. Sorry for any inconveniences.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

File: workflows/diagnostics/fv3net/diagnostics/offline/compute.py
Comment (spencerkclark) [inline workflows/diagnostics/fv3net/diagnostics/offline/compute.py:78]: I think this should fix the remaining test failures:
```suggestion
        "--catalog_path",
```
In other words we need to make sure that the `catalog_path` argument is optional for backwards compatibility (see the relevant `argparse` documentation [here](https://docs.python.org/3/library/argparse.html#name-or-flags)).  This is based on looking at this error in [the unit tests](https://app.circleci.com/pipelines/github/ai2cm/fv3net/14687/workflows/3e091ab1-73b7-40c8-b191-2c3d2fc6ab5d/jobs/139190):

<details>
<p>

```
______________ test_offline_diags_integration[multiple_datasets] _______________

data_path = '/tmp/pytest-of-root/pytest-3/test_offline_diags_integration1'

    def test_offline_diags_integration(data_path):  # noqa: F811
        """
        Test the bash endpoint for computing offline diagnostics
        """
    
        batches_kwargs = {
            "needs_grid": False,
            "res": "c8",
            "timesteps_per_batch": 1,
            "timesteps": ["20160801.001500", "20160801.003000"],
        }
        trained_model = fv3fit.testing.ConstantOutputPredictor(
            input_variables=["air_temperature", "specific_humidity"],
            output_variables=["dQ1", "dQ2"],
        )
        trained_model.set_outputs(dQ1=np.zeros([19]), dQ2=np.zeros([19]))
        data_config = loaders.BatchesFromMapperConfig(
            loaders.MapperConfig(function="open_zarr", kwargs={"data_path": data_path},),
            **batches_kwargs,
        )
        with tempfile.TemporaryDirectory() as tmpdir:
            model_dir = os.path.join(tmpdir, "trained_model")
            fv3fit.dump(trained_model, model_dir)
            data_config_filename = os.path.join(tmpdir, "data_config.yaml")
            with open(data_config_filename, "w") as f:
                yaml.safe_dump(dataclasses.asdict(data_config), f)
>           compute_diags_args = compute._get_parser().parse_args(
                [
                    model_dir,
                    data_config_filename,
                    os.path.join(tmpdir, "offline_diags"),
                    "--evaluation-grid",
                    "c8_random_values",
                    "--n-jobs",
                    "1",
                ]
            )

workflows/diagnostics/tests/offline/test_compute.py:63: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/opt/conda/envs/fv3net/lib/python3.8/argparse.py:1768: in parse_args
    args, argv = self.parse_known_args(args, namespace)
/opt/conda/envs/fv3net/lib/python3.8/argparse.py:1800: in parse_known_args
    namespace, args = self._parse_known_args(args, namespace)
/opt/conda/envs/fv3net/lib/python3.8/argparse.py:2034: in _parse_known_args
    self.error(_('the following arguments are required: %s') %
/opt/conda/envs/fv3net/lib/python3.8/argparse.py:2521: in error
    self.exit(2, _('%(prog)s: error: %(message)s\n') % args)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = ArgumentParser(prog='pytest.py', usage=None, description=None, formatter_class=<class 'argparse.HelpFormatter'>, conflict_handler='error', add_help=True)
status = 2
message = 'pytest.py: error: the following arguments are required: catalog_path\n'

    def exit(self, status=0, message=None):
        if message:
            self._print_message(message, _sys.stderr)
>       _sys.exit(status)
E       SystemExit: 2

/opt/conda/envs/fv3net/lib/python3.8/argparse.py:2508: SystemExit
```
</p>

</details>

**Review by spencerkclark — COMMENTED** [review]
@mrudko thanks for fixing the lint errors.  Sorry it looks like I missed that another test was failing.  See my comment for a suggested fix.

File: workflows/diagnostics/tests/offline/test_helpers.py
Comment (spencerkclark) [inline workflows/diagnostics/tests/offline/test_helpers.py:126]: Sorry, I did not notice this before.  As written, this is not testing any of the core code that was modified in this PR.  I think it makes more sense to call `load_grid_info` under `pytest.raises` to test that the custom catalog it gets passed is being used.  It would be great if you could add a short comment describing this too.
```suggestion
    catalog_path = os.path.join(datadir, "catalog_dummy.yaml")

    with pytest.raises(KeyError):
        load_grid_info(res, catalog_path)
```

**Review by spencerkclark — COMMENTED** [review]
Thanks @mrudko.  It looks like another test failure came up, which I think should be straightforward to fix.  See my suggestions.

<details>
<p>

```
______________ test_offline_diags_integration[multiple_datasets] _______________

data_path = '/tmp/pytest-of-root/pytest-3/test_offline_diags_integration1'

    def test_offline_diags_integration(data_path):  # noqa: F811
        """
        Test the bash endpoint for computing offline diagnostics
        """
    
        batches_kwargs = {
            "needs_grid": False,
            "res": "c8",
            "timesteps_per_batch": 1,
            "timesteps": ["20160801.001500", "20160801.003000"],
        }
        trained_model = fv3fit.testing.ConstantOutputPredictor(
            input_variables=["air_temperature", "specific_humidity"],
            output_variables=["dQ1", "dQ2"],
        )
        trained_model.set_outputs(dQ1=np.zeros([19]), dQ2=np.zeros([19]))
        data_config = loaders.BatchesFromMapperConfig(
            loaders.MapperConfig(function="open_zarr", kwargs={"data_path": data_path},),
            **batches_kwargs,
        )
        with tempfile.TemporaryDirectory() as tmpdir:
            model_dir = os.path.join(tmpdir, "trained_model")
            fv3fit.dump(trained_model, model_dir)
            data_config_filename = os.path.join(tmpdir, "data_config.yaml")
            with open(data_config_filename, "w") as f:
                yaml.safe_dump(dataclasses.asdict(data_config), f)
            compute_diags_args = compute._get_parser().parse_args(
                [
                    model_dir,
                    data_config_filename,
                    os.path.join(tmpdir, "offline_diags"),
                    "--evaluation-grid",
                    "c8_random_values",
                    "--n-jobs",
                    "1",
                ]
            )
>           compute.main(compute_diags_args)

workflows/diagnostics/tests/offline/test_compute.py:74: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
workflows/diagnostics/fv3net/diagnostics/offline/compute.py:325: in main
    evaluation_grid = load_grid_info(args.catalog_path, args.evaluation_grid)
workflows/diagnostics/fv3net/diagnostics/offline/_helpers.py:154: in load_grid_info
    if gsrm_name_from_resolution_string(res) == "scream":
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

res = '/root/project/external/vcm/vcm/catalog.yaml'

    def gsrm_name_from_resolution_string(res: str):
        """Returns the GSRM name based on the resolution"""
        if res.startswith("ne"):
            gsrm = "scream"
        elif res.startswith("c"):
            gsrm = "fv3"
        else:
>           raise ValueError(
                f"This resolution {res} can not be mapped to either scream or fv3."
            )
E           ValueError: This resolution /root/project/external/vcm/vcm/catalog.yaml can not be mapped to either scream or fv3.

external/vcm/vcm/convenience.py:158: ValueError
```

</p>
</details>

**Review by spencerkclark — COMMENTED** [review]
Perfect, thanks @mrudko -- I think this is basically ready to go now from my perspective.  I just submitted the integration tests to be sure and  I'll try and get someone else to take a quick look next week.

In the meantime could you update the PR description to remove some of the outdated items related to the `loaders` package?

**Comment by spencerkclark** [comment]
@mrudko sorry I was not able to get another pair of eyes on this this week (the best folks to review this happen to be out).  I'll see if @AnnaKwa can maybe take a look at the beginning of next week since @oliverwm1 (who we discussed this with earlier) will still be away.

File: workflows/diagnostics/fv3net/diagnostics/offline/compute.py
Comment (AnnaKwa) [inline workflows/diagnostics/fv3net/diagnostics/offline/compute.py:325]: Could you instead load the catalog just once in the `main` func (can add this above this line)? Then instead of using the catalog path as an arg to the functions `load_grid_info, get_prediciton, etc`, pass the catalog itself.

**Review by AnnaKwa — CHANGES_REQUESTED** [review]
Thanks, looks good for the most part! Just one minor request to move the catalog loading into the main function so it only has to occur once.

File: workflows/diagnostics/tests/offline/test_helpers.py
Comment (spencerkclark) [inline workflows/diagnostics/tests/offline/test_helpers.py:126]: This test needs to be updated based on the new signature of `load_grid_info`:
```suggestion
    catalog_path = os.path.join(datadir, "catalog_dummy.yaml")
    catalog = intake.open_catalog(catalog_path)

    with pytest.raises(KeyError):
        load_grid_info(catalog, res)
```

File: workflows/diagnostics/tests/offline/test_helpers.py
Comment (spencerkclark) [inline workflows/diagnostics/tests/offline/test_helpers.py:138]: This test needs to be updated based on the new signature of `load_grid_info`:
```suggestion
def test_load_grid_info(res, catalog, dim_name, expected_size):
    grid = load_grid_info(catalog, res)
```

File: workflows/diagnostics/tests/offline/test_helpers.py
Comment (spencerkclark) [inline workflows/diagnostics/tests/offline/test_helpers.py:135]:
```suggestion
    "res, catalog, dim_name, expected_size",
    [
        ("c12", vcm.catalog.catalog, "x", 12),
        ("c48", vcm.catalog.catalog, "x", 48),
        ("ne30", vcm.catalog.catalog, "ncol", 21600),
    ],
```

File: workflows/diagnostics/fv3net/diagnostics/offline/compute.py
Comment (spencerkclark) [inline workflows/diagnostics/fv3net/diagnostics/offline/compute.py:264]: I think this should be sufficient for the type hint:
```suggestion
    catalog: intake.Catalog,
```

File: workflows/diagnostics/fv3net/diagnostics/offline/compute.py
Comment (spencerkclark) [inline workflows/diagnostics/fv3net/diagnostics/offline/compute.py:287]:
```suggestion
    catalog: intake.Catalog,
```

File: workflows/diagnostics/fv3net/diagnostics/offline/_helpers.py
Comment (spencerkclark) [inline workflows/diagnostics/fv3net/diagnostics/offline/_helpers.py:15]:
```suggestion
```

File: workflows/diagnostics/fv3net/diagnostics/offline/_helpers.py
Comment (spencerkclark) [inline workflows/diagnostics/fv3net/diagnostics/offline/_helpers.py:154]:
```suggestion
def load_grid_info(catalog: intake.Catalog, res: str = "c48"):
```

**Review by spencerkclark — COMMENTED** [review]
Thanks @mrudko -- a few suggestions, some of are needed to get the tests to pass.

**Comment by AnnaKwa** [comment]
Thanks for updating the catalog arg- I believe the last failing test is due to a missing catalog arg in the test `test_load_grid_info_unknown_resolution`. 
I'll go ahead and approve since I believe the tests should all pass once you fix that.

**Review by AnnaKwa — APPROVED** [review]
Approving, just need to fix the last broken test `test_load_grid_info_unknown_resolution` by adding the missing catalog argument.

**Comment by spencerkclark** [comment]
Great, thanks @AnnaKwa -- indeed, I discussed this with @mrudko yesterday.  Hopefully he'll be able to push an update soon.

**Review by spencerkclark — APPROVED** [review]
Thanks @mrudko!  I just launched the integration tests one more time to make sure everything still passes after the recent changes.  I will merge tomorrow morning assuming all goes well there (unfortunately I cannot auto merge with the integration tests).

**Pull request merged**
